lib/std/crypto/sha2.zig

```diff
@@ -1,87 +1,114 @@
+//! Secure Hashing Algorithm 2 (SHA2)
+//!
+//! A suite of 5 secure hash functions with varying digest lengths:
+//! * Sha224
+//! * Sha256
+//! * Sha384
+//! * Sha512
+//! * Sha512/digest_len, where 1 < `digest_len` < 512 and `digest_len` != 384
+//!
+//! Published by the National Institue of Standards and Technology (NIST):
+//! https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf
+//! https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-208.pdf
+
 const std = @import("../std.zig");
 const builtin = @import("builtin");
 const mem = std.mem;
 const math = std.math;
 const htest = @import("test.zig");
 
-/////////////////////
-// Sha224 + Sha256
-
-const RoundParam256 = struct {
-    a: usize,
-    b: usize,
-    c: usize,
-    d: usize,
-    e: usize,
-    f: usize,
-    g: usize,
-    h: usize,
-    i: usize,
+pub const Sha224 = Sha2x32(iv224, 224);
+pub const Sha256 = Sha2x32(iv256, 256);
+pub const Sha384 = Sha2x64(iv384, 384);
+pub const Sha512 = Sha2x64(iv512, 512);
+
+// Truncated variants that save space but reduce security.
+// T = original IV
+// _ = changed IV
```

Comment:
Rather than talking about IVs, how about something like:

~~~
Truncating the output of SHA2-based hash functions improves security by mitigating length-extension attacks. Collision attacks remain impractical for all the types defined here.

T: original hash function, whose output is simply truncated. A truncated output is just the first bytes of a longer output.
_: hash function with context separation. Different lengths produce completely different outputs.
~~~

?

Reply:
Truncating functions based on the Merkle Damgard construction doesn't reduce security (beyond collision resistance, but anything >= 192 bits is fine), it actually improves it by keeping a large part of the state secret, mitigating length-extension attacks.

This is why SHA-512/256 is more secure than SHA-256. Unfortunately, most CPUs have acceleration for SHA-256, not SHA-512.

Reply:
This is much better than what I wrote. Updated.